### PR TITLE
[#9919] Remove HttpRequestService dependency from UploadEditProfilePictureModalComponent

### LINKED ISSUE #9919: Remove HttpRequestService dependency from UploadEditProfilePictureModalComponent
Same Reason: Reason:`UploadEditProfilePictureModalComponent` is a component and is supposed to be used in multiple places (although the reason is less convincing here as the modal is only used once. Nevertheless, let's just follow the convention). Thus, it should not be associated with a specific API request fire against the backend. All API request should be made by the corresponding pages which use it.

### PR DESCRIPTION
**Pull request by tiuweehan**
Fixes #9919

**Outline of Solution**
- Remove `HttpRequestService` dependency from `UploadEditProfilePictureModalComponent`
- Move `getProfilePicture` method from `UploadEditProfilePictureModalComponent` to `StudentProfilePageComponent`
- Create `postProfilePicture` method on `StudentProfilePageComponent`
- Refactor `onUploadEdit` to use RxJS `pipe` operator (instead of 3 layers of nested callbacks)
- Added new attribute `profilePictureEndPoint` to `StudentProfilePageComponent` to be more DRY

### CODE REVIEW

**Comment by tiuweehan** [comment]
Actually I realised that most of our endpoints are using string literals, which is not really DRY, and we may forget to make the changes if we change the endpoints on the backend. Would consolidating the endpoints into a single file (or maybe standardising the api endpoints on both frontend and backend using generators) be a valid issue to work on?

**Comment by xpdavid** [comment]
Sure. Go ahead. It is actually one of the project you can choose to work on.

Basically, we want to encapsulate API calls into service calls and use `*Request` as type in TypeScript.

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 1]
```diff
@@ -161,14 +180,28 @@ export class StudentProfilePageComponent implements OnInit {
     this.ngbModal.open(confirmDeleteProfilePicture);
   }
 
+  /**
+   * Gets the profile picture as blob image.
+   */
+  getProfilePicture(): Observable<Blob> {
+    return this.httpRequestService.get(this.profilePictureEndPoint, {}, 'blob');
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I think a better way is to create a service and encapsulate the API call. Maybe just `student-profile-service`

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 2]
```diff
@@ -161,14 +180,28 @@ export class StudentProfilePageComponent implements OnInit {
...
+  }
+
+  /**
+   * Posts the profile picture.
+   */
+  postProfilePicture(formData: FormData): Observable<any> {
+    return this.httpRequestService.post(this.profilePictureEndPoint, {}, formData);
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Same here.

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 3]
```diff
@@ -112,25 +115,41 @@ export class StudentProfilePageComponent implements OnInit {
...
-            const timestamp: number = (new Date()).getTime();
-            this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?${timestamp}&user=${this.id}`;
-          }, (response: ErrorMessageOutput) => {
-            this.statusMessageService.showErrorMessage(response.error.message);
-          });
-    }, () => {});
+    this.getProfilePicture()
+        // Return null if no picture is found
+        .pipe(
+            catchError(() => of(null)),
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I think only if the statusCode is 404 then you should return ` null`

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 4]
```diff
@@ -112,25 +115,41 @@ export class StudentProfilePageComponent implements OnInit {
...
+        )
+        // Open Modal and wait for user to upload picture
+        // Throw an error if there is no picture is uploaded
+        .pipe(
+            flatMap((image: Blob | null) => {
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Maybe should be `switchMap`

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 5]
```diff
@@ -112,25 +115,41 @@ export class StudentProfilePageComponent implements OnInit {
...
-            // force reload
-            const timestamp: number = (new Date()).getTime();
-            this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?${timestamp}&user=${this.id}`;
-          }, (response: ErrorMessageOutput) => {
-            this.statusMessageService.showErrorMessage(response.error.message);
-          });
-    }, () => {});
+    this.getProfilePicture()
+        // Return null if no picture is found
+        .pipe(
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Any reason why we don't combine pipe together into one block?
```
.pipe(
    catchError()
    flatMap()
    catchError()
    xxxx
)
```

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 6]
```diff
@@ -112,25 +115,41 @@ export class StudentProfilePageComponent implements OnInit {
...
+        // Throw an error if there is no picture is uploaded
+        .pipe(
+            flatMap((image: Blob | null) => {
+              const modalRef: NgbModalRef = this.ngbModal.open(UploadEditProfilePictureModalComponent);
+              modalRef.componentInstance.image = image;
+
+              return from(modalRef.result);
+            }),
+            catchError(() => throwError({
+              error: {
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: The error message format is for the convenient of which handler?

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 7]
```diff
@@ -112,25 +115,41 @@ export class StudentProfilePageComponent implements OnInit {
...
-            this.statusMessageService.showSuccessMessage('Your profile picture has been saved successfully');
-
-            // force reload
-            const timestamp: number = (new Date()).getTime();
-            this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?${timestamp}&user=${this.id}`;
-          }, (response: ErrorMessageOutput) => {
-            this.statusMessageService.showErrorMessage(response.error.message);
-          });
-    }, () => {});
+    this.getProfilePicture()
```
Comment (xpdavid) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Looks like you master the `rxjs` pipe. Great job!

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts (hunk 1 above)
Comment (tiuweehan) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I realize there is already  a `student-profile.service.ts` file under `services`. Should I create this method under the service?

Comment (xpdavid) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Yup.

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts (hunk 5 above)
Comment (tiuweehan) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I previously thought that catchError was a 'catch all' function for pipe, but looks like it is not the case. I will make the changes, thanks for pointing it out!

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts (hunk 6 above)
Comment (tiuweehan) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: It is for the `response` object of type `ErrorMessageOutput` in line 150

Comment (xpdavid) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: It should not be an error as users just dismiss the model?

Comment (tiuweehan) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: That makes a lot more sense, will do!

Comment (tiuweehan) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I can't find a clean way to terminate a pipe in RxJs, so in the new implementation I have passed `null` down the pipe to indicate that the user didn't upload any picture.

Would it be better to throw an error when the user does not upload any picture, but not display the error message to the user? Or is there a better alternative.

Comment (xpdavid) [reply src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: > Would it be better to throw an error when the user does not upload any picture, but not display the error message to the user?

I think so.

**Comment by tiuweehan** [comment]
@xpdavid  are you able to restart the build? The chrome e2e test is failing because they could not download nodejs, but I can't seem to find the option to restart the build.

File: src/web/services/student-profile.service.ts [hunk 8]
```diff
@@ -41,4 +42,22 @@ export class StudentProfileService {
...
+  }
+
+  /**
+   * Posts the profile picture.
+   */
+  postProfilePicture(formData: FormData): Observable<any> {
+    return this.httpRequestService.post(this.profilePictureEndPoint, {}, formData);
+  }
+
+  deleteProfilePictre(paramMap: { [key: string]: string }): Observable<any> {
```
Comment (xpdavid) [inline src/web/services/student-profile.service.ts]: Documentation?

File: src/web/services/student-profile.service.ts [hunk 9]
```diff
@@ -11,6 +11,7 @@ import { HttpRequestService } from './http-request.service';
   providedIn: 'root',
 })
 export class StudentProfileService {
+  profilePictureEndPoint: string = '/student/profilePic';
```
Comment (xpdavid) [inline src/web/services/student-profile.service.ts]: Though personally I think it is OK to use string literal here. Depends on how you want to standardize this. Either use it for all API call or not use it.

Comment (tiuweehan) [reply src/web/services/student-profile.service.ts]: Ok I'll change it back and standardize it in a later PR

**Pull request merged**